File: algorithms/Maximum Number of Occurrences of a Substring/Maximum Number of Occurrences of a Substring.cpp
class Solution {
public:
    int maxFreq(string s, int maxLetters, int minSize, int maxSize) {
        
        int i, j, k, ans = 0, res, n = s.length();
        
        string temp;
        unordered_map<char, int> st;
        unordered_map<string, int> mp;
        
        i = 0, j = 0;
        while(i <= n - minSize && j < n)
        {
            temp += s[j];
            st[s[j]] += 1;;
            if(temp.length() == minSize)
            {
                if(st.size() <= maxLetters)
                    mp[temp] += 1;
                i += 1;
                st[temp[0]] -= 1;
                if(st[temp[0]] == 0)
                    st.erase(temp[0]);
                temp.erase(temp.begin());
            }
            j += 1;
        }
        for(auto el : mp)
            ans = max(ans, el.second);
        return ans;
    }
};

File: algorithms/M/Maximum Number of Occurrences of a Substring/Maximum Number of Occurrences of a Substring.java
class Solution {
    public int maxFreq(String s, int maxLetters, int minSize, int maxSize) {
        Map<String,Integer> count = new HashMap<>();
        int ans = 0;
        int n = s.length();
        
        for(int i=0;i+minSize-1<n;i++){
            String y = s.substring(i,i+minSize);
            count.put(y,count.getOrDefault(y,0)+1);

            int unique = uniqueCharactersInString(y);
            if(unique<=maxLetters){
                ans = Math.max(ans,count.get(y));
            }
        }
        //System.out.println(count);
        return ans;
    }
    int uniqueCharactersInString(String y){
        Set<Character> hs = new HashSet<>();
        for(int j=0;j<y.length();j++) 
            hs.add(y.charAt(j));
        return hs.size();
    }
}
